src/platform/ESP32/ESP32BasicPacketFilters.h

```diff
@@ -0,0 +1,209 @@
+     * been explicitly allowed in the past. If there is no filter predicate, this is not called.
+tcarmelveilleux marked this conversation as resolved.
+     *
+     * @param endpoint - pointer to endpoint instance (platform-dependent, which is why it's void)
+     * @param pktInfo - info about source/dest of packet
+     * @param pktPayload - payload content of packet
+     */
+    virtual void OnLastMatchDequeued(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                     const chip::System::PacketBufferHandle & pktPayload)
+    {}
+
+    /**
+     * @brief Implementation of filtering before queueing that applies the predicate.
+     *
+     * See base class for arguments
+     */
+    FilterOutcome FilterBeforeEnqueue(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                      const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        // WARNING: This is likely called in a different context than `FilterAfterDequeue`. We use an atomic for the counter.
+        // Non-matching is never accounted, always allowed. Lack of predicate is equivalent to non-matching.
+        IPAddress::FromString("fe80::0000:1111:2222:3333", fakeDest);
+        IPAddress::FromString("ff02::fb", fakeMdnsDest);
+
+        if (pktInfo.DestAddress == fakeDest || pktInfo.DestAddress == fakeMdnsDest)
+        {
+            ++mNumDroppedPackets;
+            OnDropped(endpoint, pktInfo, pktPayload);
+            return FilterOutcome::kDropPacket;
+        }
+
+        if ((mPredicate == nullptr) || !mPredicate(mContext, endpoint, pktInfo, pktPayload))
+        {
+            return FilterOutcome::kAllowPacket;
+        }
+
+        if (mNumQueuedPackets.load() >= mMaxAllowedQueuedPackets)
+        {
+            ++mNumDroppedPackets;
+            OnDropped(endpoint, pktInfo, pktPayload);
+            return FilterOutcome::kDropPacket;
+        }
+
+        ++mNumQueuedPackets;
+
+        return FilterOutcome::kAllowPacket;
+    }
+
+    /**
+     * @brief Implementation of filtering after dequeueing that applies the predicate.
+     *
+     * See base class for arguments
+     */
+    FilterOutcome FilterAfterDequeue(const void * endpoint, const chip::Inet::IPPacketInfo & pktInfo,
+                                     const chip::System::PacketBufferHandle & pktPayload) override
+    {
+        // WARNING: This is likely called in a different context than `FilterBeforeEnqueue`. We use an atomic for the counter.
+        // NOTE: This is always called from Matter platform event loop
+
+        // Non-matching is never accounted, always allowed. Lack of predicate is equivalent to non-matching.
+        IPAddress::FromString("fe80::0000:1111:2222:3333", fakeDest);
+        IPAddress::FromString("ff02::fb", fakeMdnsDest);
+
+        if (pktInfo.DestAddress == fakeDest || pktInfo.DestAddress == fakeMdnsDest)
```

Comment:
This doesn't make any sense.  We dropped these so they never got enqueued, so how could they possibly be getting dequeued?